### PR TITLE
Tons of refactoring + progress bar + build ROM on background worker

### PR DESCRIPTION
**Pull request by Kjelli**
#### Tons of refactoring
- Renamed items and separated them under Items.cs (as own class and not partial)
- Decoupled ui from ROM-writing and logic building code. Populate settings from view, build rom / logic from settings.
- The settings uses enums for describing what the values are, instead of hardcoded ints.
- Item utils - some item value ranges are used in comparisons, and provide no description as to what they are. ItemUtils.cs contains utility methods to make the code more readable.
- Lots of minor readability changes like, function splitting, renaming and PascalCase convention applied.

#### Progress bar + Backgroundworker
- The UI used to block whenever a ROM was being built. The main thread was used to build the ROM, which made the application feel unresponsive. This is now happening in a background worker.
- The background worker reports progress to the progress bar, with an additional text to visualize where in the ROM-building the worker is currently on, updated real time.

Related issues:
#4

### CODE REVIEW

File: Items.cs [hunk 1]
```diff
@@ -1,418 +1,430 @@
...
+        public const int HeartPiecePicto = 52;
+        public const int HeartPieceArchB = 53;
+        public const int HeartPieceChoir = 54;
+        public const int HeartPieceBeaverRace = 55;
+        public const int HeartPieceSeaHorse = 56;
+        public const int HeartPieceFishG = 57;
+        public const int HeartPieceEvanP = 58;
+        public const int HeartPieceDogR = 59;
+        public const int HeartPiecePoeH = 60;
+        public const int HeartPieceTCGame = 61;
```
Comment (ZoeyZolotova) [inline Items.cs]: TreasureChestGame

File: Items.cs [hunk 2]
```diff
@@ -1,418 +1,430 @@
...
+        public const int HeartPiecePeahat = 62;
+        public const int HeartPieceDodong = 63;
+        public const int HeartPieceWFChest = 64;
```
Comment (ZoeyZolotova) [inline Items.cs]: WoodfallChest

File: Items.cs [hunk 3]
```diff
@@ -1,418 +1,430 @@
...
+        public const int HeartPieceTIChest = 65;
+        public const int HeartPieceOceanSH = 66;
```
Comment (ZoeyZolotova) [inline Items.cs]: OceanSpiderHouse

File: Items.cs [hunk 4]
```diff
@@ -1,418 +1,430 @@
...
+        public const int HeartPieceSeaHorse = 56;
+        public const int HeartPieceFishG = 57;
+        public const int HeartPieceEvanP = 58;
+        public const int HeartPieceDogR = 59;
+        public const int HeartPiecePoeH = 60;
+        public const int HeartPieceTCGame = 61;
+        public const int HeartPiecePeahat = 62;
+        public const int HeartPieceDodong = 63;
+        public const int HeartPieceWFChest = 64;
+        public const int HeartPieceTIChest = 65;
```
Comment (ZoeyZolotova) [inline Items.cs]: TwinIslandsChest

File: Items.cs [hunk 5]
```diff
@@ -1,418 +1,430 @@
...
-        const int ST_CLEAR = 113;
-        const int EXPLOSIVE = 114;
-        const int ARROW = 115;
-        const int WF_NEW = 116;
-        const int SH_NEW = 117;
-        const int GB_NEW = 118;
-        const int LA_NEW = 119;
-        const int IST_NEW = 120;
+        public const int AreaSouthAccess = 98;
+        public const int AreaWoodFallAccess = 99;
```
Comment (ZoeyZolotova) [inline Items.cs]: Should clarify that WoodFall / Snowhead / GreatBay / StoneTower Access/Clear refers to Temples. So rename to `AreaWoodFallTempleAccess` etc.

File: Items.cs [hunk 6]
```diff
@@ -1,418 +1,430 @@
...
+        public const int AreaSnowheadClear = 103;
+        public const int OtherEpona = 104;
+        public const int AreaWestAccess = 105;
+        public const int AreaPiratesFortressAccess = 106;
+        public const int AreaGreatBayAccess = 107;
+        public const int AreaGreatBayClear = 108;
+        public const int AreaEastAccess = 109;
+        public const int AreaIkanaCanyonAccess = 110;
+        public const int AreaStoneTowerAccess = 111;
+        public const int AreaISTAccess = 112; // ?? InSide The moon?
```
Comment (ZoeyZolotova) [inline Items.cs]: InvertedStoneTowerTemple

File: Items.cs [hunk 7]
```diff
@@ -1,418 +1,430 @@
...
+        public const int AreaStoneTowerClear = 113;
+        public const int OtherExplosive = 114;
+        public const int OtherArrow = 115;
+        public const int AreaWoodfallNew = 116;
+        public const int AreaSnowheadNew = 117;
+        public const int AreaGreatBayNew = 118;
+        public const int AreaLANew = 119; // ??
+        public const int AreaInvertedStoneTowerNew = 120; // ??
```
Comment (ZoeyZolotova) [inline Items.cs]: Correct. But these `New` ones are not used afaik

File: Items.cs [hunk 8]
```diff
@@ -1,418 +1,430 @@
...
-        const int WS_GP = 152;
-        const int BS_Bomb10 = 153;
-        const int BS_Chu10 = 154;
-        const int GS_Bomb10 = 155;
-        const int GS_Arrow10 = 156;
-        const int GS_RP = 157;
-        const int ZS_Shield = 158;
-        const int ZS_Arrow10 = 159;
-        const int ZS_RP = 160;
+        public const int ShopItemTownRedPotion = 142;
```
Comment (ZoeyZolotova) [inline Items.cs]: TP means TradingPost

File: Items.cs [hunk 9]
```diff
@@ -1,418 +1,430 @@
...
+        public const int ChestInsidePiratesFortressTankRedRupee = 179;
+        public const int ChestInsidePiratesFortressGuardSilverRupee = 180;
+        public const int ChestInsidePiratesFortressHeartPieceRoomRedRupee = 181;
+        public const int ChestInsidePiratesFortressHeartPieceRoomBlueRupee = 182;
+        public const int ChestInsidePiratesFortressMazeRedRupee = 183;
+        public const int ChestPinacleRockRedRupee1 = 184;
+        public const int ChestPinacleRockRedRupee2 = 185;
+        public const int ChestBomberHideoutSilverRupee = 186;
+        public const int GrottoTerminaBombchu = 187;
+        public const int Termina_RR_Grotto = 188; // TODO Grotto with red rupee? Grotto next to red rupee grotto?
```
Comment (ZoeyZolotova) [inline Items.cs]: To answer your question: everything in the "other chests and grotto" section except Bank_HP is in a Chest.

File: Items.cs [hunk 10]
```diff
@@ -1,418 +1,430 @@
...
+        public const int ChestTerminaUnderwaterRedRupee = 189;
+        public const int ChestTerminaGrassRedRupee = 190;
+        public const int ChestTerminaStumpRedRupee = 191;
+        public const int GrottoGreatBayCoast = 192;
+        public const int GB_Cape_Ledge_1 = 193; // TODO Grotto or chest?
```
Comment (ZoeyZolotova) [inline Items.cs]: Chest. ledges to get to the beavers

File: Items.cs [hunk 11]
```diff
@@ -1,418 +1,430 @@
...
+        public const int GB_Cape_Ledge_2 = 194; // TODO Grotto or chest?
+        public const int GrottoGreatBayCape = 195;
+        public const int GB_Cape_Underwater = 196; // TODO Grotto or chest?
```
Comment (ZoeyZolotova) [inline Items.cs]: Chest

File: Items.cs (hunk 9 above)
Comment (ZoeyZolotova) [reply Items.cs]: Are you trying to separate which chests are in grottos and which aren't?

File: Items.cs [hunk 12]
```diff
@@ -1,418 +1,430 @@
...
+        public const int HeartPieceNorthClockTown = 230;
+        public const int HeartPieceToSwamp = 231;
+        public const int HeartPieceSwampScrub = 232;
+        public const int HeartPieceDekuPalace = 233;
+        public const int HeartPieceGoronVillageScrub = 234;
+        public const int HeartPieceZoraGrotto = 235;
+        public const int HeartPieceLabFish = 236;
+        public const int HeartPieceGreatBayCapeLikeLike = 237;
+        public const int HeartPiecePiratesFortress = 238;
+        public const int HeartPieceZoraHall = 239;
```
Comment (ZoeyZolotova) [inline Items.cs]: ZoraHallScrub

File: Items.cs [hunk 13]
```diff
@@ -1,418 +1,430 @@
...
+        public const int ItemTingleMapTown = 248;
+        public const int ItemTingleMapWoodfall = 249;
+        public const int ItemTingleMapSnowhead = 250;
+        public const int ItemTingleMapRanch = 251;
+        public const int ItemTingleMapGreatBay = 252;
+        public const int ItemTingleMapStoneTower = 253;
+
         //oops I forgot one
-        const int To_GR_Grotto = 254;
+        public const int GrottoToGR = 254;
```
Comment (ZoeyZolotova) [inline Items.cs]: GrottoToGoronRace

File: Majora's Mask Randomiser.csproj [hunk 14]
```diff
@@ -49,7 +49,8 @@
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>
   <PropertyGroup>
-    <ApplicationIcon>200px_majoramask_rtw_icon.ico</ApplicationIcon>
+    <ApplicationIcon>
+    </ApplicationIcon>
```
Comment (ZoeyZolotova) [inline Majora's Mask Randomiser.csproj]: Revert this change please.

File: Majora's Mask Randomiser.csproj [hunk 15]
```diff
@@ -741,7 +750,6 @@
     </BootstrapperPackage>
   </ItemGroup>
   <ItemGroup>
-    <Content Include="200px_majoramask_rtw_icon.ico" />
```
Comment (ZoeyZolotova) [inline Majora's Mask Randomiser.csproj]: Revert

File: Shuffle.cs [hunk 16]
```diff
@@ -1,345 +1,469 @@
...
         {
             // Gold Dust cannot be obtained a second time at certain locations
             // All chests are Recovery Heart the 2nd time
             {
-                Bottle_G, new List<int>
+                Items.ItemBottleGoronRace, new List<int>
                 {
-                    Hero_Bow, Fire_Arrow, Ice_Arrow, Light_Arrow, Lens, Hookshot, Bottle_D, M_Shield,
-                    HP_Postman, // Rewards 50 rupees when collecting 2nd time
-                    HP_Keaton, HP_SwSch, HP_DogR, // Rewards 20 rupees when collecting 2nd time
```
Comment (ZoeyZolotova) [inline Shuffle.cs]: Looks like you missed DogRace when you resolved conflicts.

File: Shuffle.cs [hunk 17]
```diff
@@ -892,230 +1054,310 @@ private bool CheckMatch(int CurrentItem, int Target)
...
+                        free = RNG.Next(Items.SongOath + 1);
                     }
                 }
-                ItemList[free].Replaces = Deku_Mask;
-                TargetPool.Remove(Deku_Mask);
+                ItemList[free].ReplacesItemId = Items.MaskDeku;
+                ItemPool.Remove(Items.MaskDeku);
             };
-            for (int i = Deku_Mask; i < HP_Mayor; i++)
+            for (int i = Items.MaskDeku; i < Items.HeartPieceNotebookMayor; i++)
```
Comment (ZoeyZolotova) [inline Shuffle.cs]: Since you're changing these, can you change the condition to be `<=` instead of `<`? Much more understandable that way.
i.e. `for (int i = Items.MaskDeku; i <= Items.TradeItemMamaLetter; i++)`
Same for the ones below.

File: Items.cs [hunk 18]
```diff
@@ -1,418 +1,430 @@
...
+        public const int HeartPieceGosTF = 49;
+        public const int HeartPieceScrTF = 50;
+        public const int HeartPieceArchS = 51;
+        public const int HeartPiecePicto = 52;
+        public const int HeartPieceArchB = 53;
+        public const int HeartPieceChoir = 54;
+        public const int HeartPieceBeaverRace = 55;
+        public const int HeartPieceSeaHorse = 56;
+        public const int HeartPieceFishG = 57;
+        public const int HeartPieceEvanP = 58;
```
[the remark this answers is not in the document]
Comment (Kjelli) [reply Items.cs]: Can't see what the P would mean, changing to `Evan`.

File: Items.cs (hunk 9 above)
Comment (Kjelli) [reply Items.cs]: Oh, I see - thanks. Nah, I think I tried to derive the type of "item" it was from its previous name, when it didn't include chest or hp, I assumed it was specifying an actual grotto. Will include chest in the name for clarification.

File: Items.cs [hunk 19]
```diff
@@ -1,418 +1,430 @@
...
+        public const int ChestInvertedStoneTowerBombchu10 = 213;
+        public const int ChestInvertedStoneTowerBean = 214;
+        public const int ChestToSnowheadGrotto = 215; 
+        public const int ChestToGoronVillageRedRupee = 216; 
+        public const int ChestSecretShrineHeartPiece = 217; 
+        public const int ChestSecretShrineDinoGrotto = 218; 
+        public const int ChestSecretShrineWizzGrotto = 219; 
+        public const int ChestSecretShrineWartGrotto = 220; 
+        public const int ChestSecretShrineGaroGrotto = 221; 
+        public const int ChestInnStaffRoomGrotto = 222; // + are these two grottos?
```
Comment (ZoeyZolotova) [inline Items.cs]: These are not grottos. Inn Staff Room is the room where you get the couple's mask on the last night. There's a chest in there. Inn Guest Room is the room you can open with the Room Key.

File: Items.cs [hunk 20]
```diff
@@ -1,418 +1,430 @@
...
+        public const int ItemTingleMapTown = 248;
+        public const int ItemTingleMapWoodfall = 249;
+        public const int ItemTingleMapSnowhead = 250;
+        public const int ItemTingleMapRanch = 251;
+        public const int ItemTingleMapGreatBay = 252;
+        public const int ItemTingleMapStoneTower = 253;
+
         //oops I forgot one
-        const int To_GR_Grotto = 254;
+        public const int GrottoToGoronRace = 254;
```
Comment (ZoeyZolotova) [inline Items.cs]: Should be ChestToGoronRaceGrotto to be consistent.

File: Items.cs [hunk 21]
```diff
@@ -1,418 +1,430 @@
...
+            ShopItemGoronArrow10,
+            ShopItemBombsBomb10,
+            ShopItemBombsBombchu10,
+            ShopItemZoraArrow10,
+            ShopItemZoraRedPotion,
+            ShopItemZoraShield,
         };
 
-        int[] ITEM_ADDRS = new int[]
+        public static readonly int[] ITEM_ADDRS = new int[]
```
Comment (ZoeyZolotova) [inline Items.cs]: These should be ReadOnlyCollection i think.

File: Items.cs (hunk 19 above)
Comment (Kjelli) [reply Items.cs]: Okay, I was thinking the same thing - but for some reason they were named grottos originally.
